drivers/nvram_spi/nvram-spi.c

```diff
@@ -0,0 +1,284 @@
+    int status;
+    union {
+        uint32_t u32;
+        char c[4];
+    } addr;
+    /* Address is expected by the device as big-endian, i.e. network byte order,
+     * we utilize the network byte order macros here. */
+    addr.u32 = HTONL(dst);
+    /* Acquire exclusive bus access */
+    spi_acquire(spi_dev->spi);
+    /* Assert CS */
+    gpio_clear(spi_dev->cs);
+    /* Enable writes */
+    status = spi_transfer_byte(spi_dev->spi, NVRAM_SPI_CMD_WREN, NULL);
+    if (status < 0)
+    {
+        return status;
+    }
+    /* Release CS */
+    gpio_set(spi_dev->cs);
+    hwtimer_spin(NVRAM_SPI_CS_TOGGLE_TICKS);
+    /* Re-assert CS */
+    gpio_clear(spi_dev->cs);
+    /* Write command and address */
+    status = spi_transfer_regs(spi_dev->spi, NVRAM_SPI_CMD_WRITE,
+                      &addr.c[sizeof(addr.c) - spi_dev->address_count], NULL,
+                      spi_dev->address_count);
+    if (status < 0)
+    {
+        return status;
+    }
+    /* Keep holding CS and write data */
+    status = spi_transfer_bytes(spi_dev->spi, (char *)src, NULL, len);
+    if (status < 0)
+    {
+        return status;
+    }
+    /* Release CS */
+    gpio_set(spi_dev->cs);
+    /* Release exclusive bus access */
+    spi_release(spi_dev->spi);
+    return status;
+}
+
+static int nvram_spi_read(nvram_t *dev, uint8_t *dst, uint32_t src, size_t len)
+{
+    nvram_spi_params_t *spi_dev = (nvram_spi_params_t *) dev->extra;
+    int status;
+    union {
+        uint32_t u32;
+        char c[4];
+    } addr;
+    /* Address is expected by the device as big-endian, i.e. network byte order,
+     * we utilize the network byte order macros here. */
+    addr.u32 = HTONL(src);
+    /* Acquire exclusive bus access */
+    spi_acquire(spi_dev->spi);
+    /* Assert CS */
+    gpio_clear(spi_dev->cs);
+    /* Write command and address */
+    status = spi_transfer_regs(spi_dev->spi, NVRAM_SPI_CMD_READ,
+                               &addr.c[sizeof(addr.c) - spi_dev->address_count],
+                               NULL, spi_dev->address_count);
+    if (status < 0)
+    {
+        return status;
+    }
+    /* Keep holding CS and read data */
+    status = spi_transfer_bytes(spi_dev->spi, NULL, (char *)dst, len);
+    if (status < 0)
+    {
+        return status;
+    }
+    /* Release CS */
+    gpio_set(spi_dev->cs);
+    /* Release exclusive bus access */
+    spi_release(spi_dev->spi);
+    /* status contains the number of bytes actually read from the SPI bus. */
+    return status;
+}
+
+
+static int nvram_spi_write_9bit_addr(nvram_t *dev, uint8_t *src, uint32_t dst, size_t len)
+{
+    nvram_spi_params_t *spi_dev = (nvram_spi_params_t *) dev->extra;
+    int status;
+    uint8_t cmd;
+    uint8_t addr;
+    cmd = NVRAM_SPI_CMD_WRITE;
+    /* The upper address bit is mixed into the command byte on certain devices,
+     * probably just to save a byte in the SPI transfer protocol. */
+    if (dst > 0xff) {
+        cmd |= 0x08;
```

Comment:
Ah, the `cmd` just uses the lower three bits and therefore the 9th address bit is set at the 4th position?